File: 01-programming-languages/JavaScript/01-fundamentals/02-operators/operators.js
/*
 * JavaScript Operators
 * Arithmetic, comparison, logical, nullish
 */

// ============================================
// Arithmetic Operators
// ============================================

const a = 10;
const b = 3;

console.log("Arithmetic Operators:");
console.log(`a + b = ${a + b}`);   // Addition: 13
console.log(`a - b = ${a - b}`);   // Subtraction: 7
console.log(`a * b = ${a * b}`);   // Multiplication: 30
console.log(`a / b = ${a / b}`);   // Division: 3.333...
console.log(`a % b = ${a % b}`);   // Modulo: 1
console.log(`a ** b = ${a ** b}`); // Exponentiation: 1000

// Increment and Decrement
let counter = 5;
counter++; // Post-increment: 6
++counter; // Pre-increment: 7
counter--; // Post-decrement: 6
--counter; // Pre-decrement: 5

// ============================================
// Comparison Operators
// ============================================

console.log("\nComparison Operators:");

// Loose equality (==) - performs type coercion
console.log(5 == "5");    // true
console.log(0 == false);  // true
console.log(null == undefined); // true

// Strict equality (===) - no type coercion
console.log(5 === "5");   // false
console.log(0 === false); // false
console.log(null === undefined); // false

// Inequality operators
console.log(5 != "5");    // false (loose)
console.log(5 !== "5");   // true (strict)
console.log(5 > 3);      // true
console.log(5 < 3);      // false
console.log(5 >= 5);     // true

// ============================================
// Logical Operators
// ============================================

console.log("\nLogical Operators:");

const x = true;
const y = false;

// AND (&&) - returns first falsy or last value
console.log(x && y);       // false
console.log(x && "hello"); // "hello"
console.log(y && "hello"); // false

// OR (||) - returns first truthy or last value
console.log(x || y);            // true
console.log(y || "hello");      // "hello"
console.log("" || "default");   // "default"

// NOT (!) - converts to boolean then negates
console.log(!true);  // false
console.log(!0);     // true
console.log(!"");    // true

// ============================================
// Nullish Coalescing (??)
// ============================================

// Returns right side only if left is null or undefined
console.log(null ?? "default");       // "default"
console.log(undefined ?? "default");  // "default"
console.log(0 ?? "default");         // 0 (0 is not nullish)
console.log("" ?? "default");        // "" (empty string is not nullish)

// ============================================
// Nullish Coalescing Assignment (??=)
// ============================================

let config = null;
config ??= { theme: "dark" }; // Assigns if null/undefined
console.log(config); // { theme: "dark" }

let existing = "keep";
existing ??= "new"; // Does not assign
console.log(existing); // "keep"

// ============================================
// Optional Chaining (?.)
// ============================================

const user = {
    name: "Alice",
    address: {
        city: "Wonderland"
    }
};

console.log(user?.name);          // "Alice"
console.log(user?.address?.city); // "Wonderland"
console.log(user?.phone);         // undefined (no error)

// ============================================
// Ternary Operator
// ============================================

const age = 20;
const status = age >= 18 ? "adult" : "minor";
console.log(`Status: ${status}`); // "adult"

// ============================================
// Operator Precedence
// ============================================

// Remember: PEMDAS/BODMAS applies
console.log(2 + 3 * 4);   // 14 (not 20)
console.log((2 + 3) * 4); // 20

// Best Practice: Use parentheses for clarity
const result = (a + b) * (a - b);

File: 01-programming-languages/C#/01-fundamentals/02-operators/operators.cs
// C# Operators

using System;

namespace OperatorsDemo
{
    class Program
    {
        static void Main(string[] args)
        {
            // Arithmetic Operators
            int a = 10, b = 3;
            Console.WriteLine($"Addition: {a + b}");      // 13
            Console.WriteLine($"Subtraction: {a - b}");   // 7
            Console.WriteLine($"Multiplication: {a * b}"); // 30
            Console.WriteLine($"Division: {a / b}");      // 3 (integer division)
            Console.WriteLine($"Remainder: {a % b}");     // 1
            
            // Floating point division
            double x = 10.0, y = 3.0;
            Console.WriteLine($"Float Division: {x / y}"); // 3.333...
            
            // Increment/Decrement
            int counter = 5;
            counter++;
            Console.WriteLine($"After increment: {counter}"); // 6
            counter--;
            Console.WriteLine($"After decrement: {counter}"); // 5
            
            // Comparison Operators
            Console.WriteLine($"Equal: {a == b}");         // False
            Console.WriteLine($"Not Equal: {a != b}");     // True
            Console.WriteLine($"Greater: {a > b}");        // True
            Console.WriteLine($"Less: {a < b}");           // False
            Console.WriteLine($"Greater/Equal: {a >= b}"); // True
            Console.WriteLine($"Less/Equal: {a <= b}");    // False
            
            // Logical Operators
            bool p = true, q = false;
            Console.WriteLine($"AND: {p && q}");           // False
            Console.WriteLine($"OR: {p || q}");            // True
            Console.WriteLine($"NOT: {!p}");               // False
            
            // Bitwise Operators
            int m = 12, n = 10; // 1100 & 1010
            Console.WriteLine($"AND: {m & n}");            // 8 (1000)
            Console.WriteLine($"OR: {m | n}");             // 14 (1110)
            Console.WriteLine($"XOR: {m ^ n}");            // 6 (0110)
            Console.WriteLine($"Left Shift: {m << 2}");    // 48
            Console.WriteLine($"Right Shift: {m >> 2}");   // 3
            
            // Assignment Operators
            int val = 10;
            val += 5;   // val = val + 5
            val -= 3;   // val = val - 3
            val *= 2;   // val = val * 2
            val /= 4;   // val = val / 4
            
            // Ternary Operator
            int age = 20;
            string status = age >= 18 ? "Adult" : "Minor";
            Console.WriteLine($"Status: {status}");
            
            // Null-coalescing Operator
            string? nullStr = null;
            string result = nullStr ?? "default";
            Console.WriteLine($"Result: {result}");
            
            // Null-conditional Operator
            string? name = null;
            int? length = name?.Length;
            Console.WriteLine($"Length: {length ?? 0}");
            
            // Type checking Operators
            object obj = "hello";
            Console.WriteLine($"Is string: {obj is string}");
            Console.WriteLine($"As string: {(obj as string)?.ToUpper()}");
            
            Console.WriteLine("Operators example running");
        }
    }
}